clients/src/main/java/org/apache/kafka/common/record/Records.java

```diff
@@ -90,4 +91,18 @@ public interface Records extends TransferableRecords {
      * @return The record iterator
      */
     Iterable<Record> records();
+
+    /**
+     * Return a slice of records from this instance, which is a view into this set starting from the given position
+     * and with the given size limit.
+     *
+     * If the size is beyond the end of the records, the end will be based on the size of the records at the time of the read.
+     *
+     * If this records set is already sliced, the position will be taken relative to that slicing.
+     *
+     * @param position The start position to begin the read from
+     * @param size The number of bytes after the start position to include
+     * @return A sliced wrapper on this message set limited based on the given position and size
+     */
+    Records slice(int position, int size) throws IOException;
```

Comment:
This part doesn't seem right to me. Neither `MemoryRecords` nor `FileRecords` implements a slice method which returns `Records`. There would be a bit more work required to make the concrete classes properly interchangeable so the caller didn't need to know which class they were dealing with.

Reply:
Could we have both MemoryRecords and FileRecords return Records?

Reply:
Yes, we could, but there would be downstream effects of that.

Reply:
I have made a separate [commit](https://github.com/apache/kafka/pull/19581/commits/7eb661be549a31ba2339ee556cc313b96679c482) in this PR which shows the changed to make MemoryRecords and FileRecords to return Record type. The casting is required at 1 place in DumpLogSegments.scala. Couple of cast is required in test file where I am validating `validBytes` in MemoryRecordTest. Otherwise the change looks good.

Reply:
Thanks for he explanation. We can just keep what's in the PR then.